File: 32-longest-valid-parentheses/32-longest-valid-parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        stack = []
        res = []
        temp = 0
        for char in s:
            if char == "(":
                res.append(temp)
                temp = 0
                stack.append(char)
            else:
                if stack and stack[-1] == "(":
                    temp += res.pop() + 1
                    stack.pop()
                else:
                    res.append(temp)
                    temp = 0
        res.append(temp)
        return max(res)*2
                
                    
                    
        

File: 32-longest-valid-parentheses/32-longest-valid-parentheses.java
class Solution {
    public int longestValidParentheses(String s) {
        int rollingLeft = 0, rollingRight = 0;
        int ans = 0;
        //left-pass
        for(int i = 0; i < s.length(); i++){
            char curChar = s.charAt(i);
            if(curChar == '('){
                rollingLeft += 1;
            
            }else{
                rollingRight += 1;
            }
            if(rollingLeft == rollingRight){
                ans = Math.max(ans, rollingRight * 2);
                
            }else if(rollingLeft < rollingRight){
                rollingLeft = rollingRight = 0;
            }
        }
        rollingLeft = 0;
        rollingRight = 0;
        for(int i = s.length() - 1; i >= 0; i--){
            char curChar = s.charAt(i);
            if(curChar == '('){
                rollingLeft += 1;
            
            }else{
                rollingRight += 1;
            }
            if(rollingLeft == rollingRight){
                ans = Math.max(ans, 2 * rollingRight);
                
            }else if(rollingLeft > rollingRight){
                rollingLeft = rollingRight = 0;
            }
        }
        return ans;
        
    
    }

}